kong/tools/public.lua

```diff
@@ -12,62 +12,139 @@ do
 
 
   local str_find              = string.find
+  local str_format            = string.format
   local ngx_req_get_post_args = ngx.req.get_post_args
   local ngx_req_get_body_data = ngx.req.get_body_data
 
 
-  local content_type_map = {
-    [1] = function(content_type) -- multipart
-      return multipart(ngx_req_get_body_data(), content_type):get_all()
+  local MIME_TYPES = {
+    form_url_encoded = 1,
+    json             = 2,
+    xml              = 3,
+    multipart        = 4,
+    text             = 5,
+    html             = 6,
+  }
+
+
+  local ERRORS     = {
+    no_ct          = 1,
+    [1]            = "don't know how to parse request body: no Content-Type",
+    unknown_ct     = 2,
+    [2]            = "don't know how to parse request body: " ..
+                     "unknown Content-Type: '%s'",
+    unsupported_ct = 3,
+    [3]            = "don't know how to parse request body: " ..
+                     "can't decode Content-Type: '%s'"
+  }
+
+
+  _M.req_mime_types  = MIME_TYPES
+  _M.req_body_errors = ERRORS
+
+
+  local MIME_DECODERS = {
+    [MIME_TYPES.multipart] = function(content_type)
+      local raw_body = ngx_req_get_body_data()
+      local args     = multipart(raw_body, content_type):get_all()
+
+      return args, raw_body
     end,
 
-    [2] = function() -- json
-      local body, err = cjson.decode(ngx_req_get_body_data())
+    [MIME_TYPES.json] = function()
+      local raw_body  = ngx_req_get_body_data()
+      local args, err = cjson.decode(raw_body)
       if err then
         ngx_log(ERR, "could not decode JSON body args: ", err)
-        return {}
+        return {}, raw_body
       end
 
-      return body
+      return args, raw_body
     end,
 
-    [3] = function() -- encoded form
+    [MIME_TYPES.form_url_encoded] = function()
       local ok, res, err = pcall(ngx_req_get_post_args)
       if not ok or err then
         local msg = res and res or err
         ngx_log(ERR, "could not get body args: ", msg)
         return {}
       end
 
+      -- don't read raw_body if not necessary
+      -- if we called get_body_args(), we only want the parsed body
       return res
     end,
   }
 
-  function _M.get_body_args()
+
+  local function get_body_info()
     local content_type = ngx.var.http_content_type
```

Comment:
this doesnt seem to handle multiple Content-Type headers, which something i should have done when initially writing this. not technically in the scope of this change, but probably something we should handle

Reply:
Not tackling this here, but keeping track of it you know where.